### DISCUSSION TITLE
How to / When to use VmDatabase.opened(..) ?

### DISCUSSION THREAD

**Discussion by wph144** [simolus3/drift#1403] [Q&A]
Although I've read comments on VmDatabase.opened, I did not get it.
Can you explain when it is useful? I wonder that about closeUnderlyingOnClose too.
Thanks!

**Comment by simolus3** [comment]
The unnamed and the `.memory` constructor on `VmDatabase` will open a new sqlite3 database when used. On the other hand, `VmDatabase.opened` will use a database connection that is already open. Further, closing a `VmDatabase` (which happens indirectly by closing the database using it), will close the underlying connection as well.  When `closeUnderlyingOnClose` is set to `false`, it won't.

This is mainly used by moor internally, for instance when [verifying schema migrations](https://moor.simonbinder.eu/docs/advanced-features/migrations/#verifying-migrations). The implementation basically:

1. Uses the `sqlite3` package to create a named in-memory database
2. Then uses `VmDatabase.opened` multiple times to setup the initial schema and connections used by user code to call their migrations
3. Uses the raw database to compare the schema with what's expected.
